returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,205 @@
+            results are passed to the iterator-level function.
+        :param output: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        assert (
+            self.seq_ordering == "default"
+        ), f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}"
+        assert (
+            map_seq is not None or map_seq_stream is not None
+        ), "need to either define map_seq or map_seq_stream functions"
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq or self._identity
+        self._map_seq_stream = map_seq_stream or self._identity
+        self._map_outputs = map_outputs
+
+        self._dataset: Optional[Dataset] = None
+        self._data_keys: Optional[List[str]] = None
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._data_iter_seq_idx = -1
+        self._dim_cache: Dict[str, List[Dim]] = {}
+
+        if _meta_info_cache:
+            self._dim_cache = _meta_info_cache["_dim_cache"]
+            self._estimated_num_seqs = _meta_info_cache["_estimated_num_seqs"]
+            self.labels = _meta_info_cache["labels"]
+            self.num_inputs = _meta_info_cache["num_inputs"]
+            self.num_outputs = _meta_info_cache["num_outputs"]
+
+    def initialize(self):
+        """init"""
+        self._lazy_init_num_outputs()
+        super().initialize()
+
+    @property
+    def _meta_info_cache(self) -> Optional[Dict[str, Any]]:
+        if self.num_outputs is None:
+            return None
+        return {
+            "_dim_cache": self._dim_cache,
+            "_estimated_num_seqs": self._estimated_num_seqs,
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
+
+    def _lazy_init_num_outputs(self):
+        if self.num_outputs is not None:
+            return
+
+        dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self.labels = dataset.labels
+        self._estimated_num_seqs = dataset.estimated_num_seqs
+        if self._map_outputs is not None:
+            tdict = TensorDict(self._map_outputs)
+            # TODO: What if feature_dim == None?
+            self.num_inputs = tdict.data["data"].feature_dim.size
+            self.num_outputs = {
+                k: (t.sparse_dim.size, 1) if t.sparse_dim is not None else (t.feature_dim.size, t.ndim)
+                for k, t in tdict.data.items()
+            }
```

Comment:
```suggestion
            self.num_outputs = {
                k: (t.sparse_dim.size if t.sparse_dim else t.shape[-1] or 1, t.ndim)
                for k, t in tdict.data.items()
            }
            self.num_inputs = self.num_outputs[self._default_input][0]
```

Reply:
You must also derive `self._default_input`. It's ok to use some heuristics, it shouldn't matter.

Reply:
Btw, see `AnythingDataset`. There I do this:
```python
def __init__(self, *, data_keys: Dict[str, Dict[str, Any]], **kwargs):
    """
    :param data_keys: similar like extern_data. defines shape, dtype, sparse, dim, etc
    """
    super().__init__(**kwargs)
    self._data_keys = data_keys
    self._input_key = next(iter(data_keys))
    self.num_inputs = data_keys[self._input_key]["dim"]
    self.num_outputs = {k: (v["dim"], len(v["shape"])) for (k, v) in data_keys.items()}
```